src/main/scala/cromwell/binding/Task.scala

```diff
@@ -52,27 +31,121 @@ object Task {
  * Represents a `task` declaration in a WDL file
  *
  * @param name Name of the task
- * @param command Abstract command defined in the `command` section
+ * @param declarations Any declarations (e.g. String something = "hello") defined in the task
+ * @param commandTemplate Sequence of command pieces, essentially a parsed command template
  * @param outputs Set of defined outputs in the `output` section of the task
+ * @param ast The syntax tree from which this was built.
+ * @param backendType which backend this task is meant to run on.
  */
 case class Task(name: String,
                 declarations: Seq[Declaration],
-                command: Command,
+                commandTemplate: Seq[CommandPart],
                 outputs: Seq[TaskOutput],
                 ast: Ast,
                 backendType: BackendType) extends Executable {
+  val ws = Pattern.compile("[\\ \\t]+")
+
+  /**
+   * Attributes defined in the runtime {...} section of a WDL task
+   */
+  val runtimeAttributes = RuntimeAttributes(ast, backendType)
+
+  /**
+   * Inputs to this task, as locally qualified names
+   *
+   * @return Seq[TaskInput] where TaskInput contains the input
+   *         name & type as well as any postfix quantifiers (?, +)
+   */
+  val inputs: Seq[TaskInput] = for(declaration <- declarations; input <- declaration.asTaskInput) yield input
+  // TODO: I think TaskInput can be replaced by Declaration
+
+  /**
+   * Given a map of task-local parameter names and WdlValues, create a command String.
+   *
+   * Instantiating a command line is the process of taking a command in this form:
+   *
+   * {{{
+   *   sh script.sh ${var1} -o ${var2}
+   * }}}
+   *
+   * This command is stored as a `Seq[CommandPart]` in the `Command` class (e.g. [sh script.sh, ${var1}, -o, ${var2}]).
+   * Then, given a map of variable -> value:
+   *
+   * {{{
+   * {
+   *   "var1": "foo",
+   *   "var2": "bar"
+   * }
+   * }}}
+   *
+   * It calls instantiate() on each part, and passes this map. The ParameterCommandPart are the ${var1} and ${var2}
+   * pieces and they lookup var1 and var2 in that map.
+   *
+   * The command that's returned from Command.instantiate() is:
+   *
+   *
+   * {{{sh script.sh foo -o bar}}}
+   *
+   * @param parameters Parameter values
+   * @return String instantiation of the command
+   */
+  def instantiateCommand(parameters: CallInputs, functions: WdlFunctions = new NoFunctions): Try[String] = {
+    Try(normalize(commandTemplate.map { _.instantiate(declarations, parameters, functions) }.mkString("")))
+  }
+
+  def commandTemplateString: String = normalize(commandTemplate.map(_.toString).mkString)
```

Comment:
Could we (likely via an implicit value class) make normalize the method so it looks like a chained operation instead of a function call?